operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,307 @@
+
+	// Reconcile current state with desired state at every
+	// ciliumEndpointBatchUpsertSyncPeriod
+	log.WithFields(logrus.Fields{
+		"Upsert syncperiod": c.ciliumEndpointBatchUpsertSyncPeriod,
+		"Delete syncperiod": c.ciliumEndpointBatchDeleteSyncPeriod,
+	}).Debug("Starting CEB controller reconciler.")
+
+	go wait.Until(c.worker, c.ciliumEndpointBatchUpsertSyncPeriod, stopCh)
+
+	go func() {
+		defer utilruntime.HandleCrash()
+	}()
+
+	<-stopCh
+
+	return
+}
+
+// Upon warm boot[restart], Iterate over all CEPs which we got from the api-server
+// and compare it with CEPs packed inside CEB.
+// If there are any stale CEPs present in CEBs, remove them from their CEB.
+func (c *CiliumEndpointBatchController) ciliumRemoveStaleCepEntries() {
+	log.Info("Remove stale CEP entries in CEB")
+
+	// Get all ceps from local datastore
+	staleCeps := c.Manager.getAllCeps()
+
+	// Remove stale CEP entries present in CEB
+	for cepName, cCep := range staleCeps {
+		if _, exists, err := c.ciliumEndpointStore.GetByKey(getCepNameFromCCEP(cCep)); err == nil && exists || err != nil {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"cep-name": cepName,
+		}).Debug("Removing stale CEP entry.")
+		c.Manager.RemoveCepFromCache(cCep)
+	}
+	return
+}
+
+// syncs all cebs from cebStore in manager cache.
+func syncCebsInLocalCache(cebStore cache.Store, manager cebManager) {
+
+	for _, obj := range cebStore.List() {
+		ceb := obj.(*v2alpha1.CiliumEndpointBatch)
+		// If CEB is already cached locally, do nothing.
+		if _, err := manager.getCebFromCache(ceb.GetName()); err == nil {
+			continue
+		}
+
+		// Create new CEB locally, with the given cebName
+		manager.createCeb(ceb.GetName())
+
+		// Deep copy the ceb, we got from api-server to local datastore.
+		manager.updateCebInCache(ceb, true)
+
+	}
+	log.Debug("Successfully synced all CEBs locally")
+	return
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same ceb
+// at the same time
+func (c *CiliumEndpointBatchController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CiliumEndpointBatchController) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncCeb(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *CiliumEndpointBatchController) handleErr(err error, key interface{}) {
+
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+}
+
+// syncCeb reconciles the queued CEB with api-server.
+func (c *CiliumEndpointBatchController) syncCeb(key string) error {
+
+	var empty string
+	// Check for valid CEB in local cache, if not return error
+	if _, er := c.Manager.getCebFromCache(key); er != nil {
+		return er
```

Comment:
One scenario is CEB delete succeeded, but somehow we got update request for the `ceb` .

Reply:
There is always possibility that, in multi threaded environment we can expect that some events are out of order.

Reply:
We need to be certain, otherwise you'd need to add guards EVERYWHERE.

Reply:
I wish i know all corner cases, issues, race conditions so we can happily run with less code without any guardrails.  The reason for these checks is to log these failures, thats helps in debug later point.

Reply:
My point is that this could be a bug instead of an expected scenario.

Reply:
Checking for valid object is the right thing to do. Whole point of having these check is to uncover the bugs `gracefully`.

Reply:
I beg to disagree. Unless you know that this condition will happen, otherwise this is just hiding the bug away, because you shouldn't hit this condition in the first place...@varunmar PTAL.

Reply:
I'm removing the check for invalid object.